lib/sms/sms_submit.c

```diff
@@ -88,6 +103,164 @@ static int sms_submit_encode_number(
+	 * and SCA if the <response> is longer than 255 bytes (510 characters).
+	 * This limitation should not be a problem in practice.
+	 */
+
+	/* Check if this is a type approval SIM */
+
+	/* Read EFAD (0x6FAD = 28589): 3GPP TS 51.011 chapter 10.3.18 */
+	ret = nrf_modem_at_scanf(
+		"AT+CRSM=176,28589,0,0,0",
+		"+CRSM: %d, %d, \"%511[^\"]\"",
+		&crsm_sw1,
+		&crsm_sw2,
+		sms_buf_tmp);
+
+	LOG_DBG("EFAD: (ret=%d, crsm_sw1=%d, crsm_sw2=%d): %s",
+		ret, crsm_sw1, crsm_sw2, sms_buf_tmp);
+
+	if (ret != 3 || crsm_sw1 != 0x90 || crsm_sw2 != 0x00 || strlen(sms_buf_tmp) == 0) {
+		LOG_WRN("Failed to read UICC EFSMSP (ret=%d, crsm_sw1=%d, crsm_sw2=%d): %s",
+			ret, crsm_sw1, crsm_sw2, sms_buf_tmp);
+		goto no_ta_sim;
+	}
+	if (sms_buf_tmp[0] != '8') {
+		LOG_DBG("Not a type approval SIM");
+		goto no_ta_sim;
+	}
+
+	LOG_DBG("Type approval SIM detected");
+
+	/* Check if SIM has SMS Service Center Address */
+	sms_buf_tmp[0] = '\0';
+
+	/* Read EFSMSP (0x6F42 = 28482): 3GPP TS 51.011 chapter 10.5.6 */
+	ret = nrf_modem_at_scanf(
+		"AT+CRSM=178,28482,1,4,0",
+		"+CRSM: %d, %d, \"%511[^\"]\"",
+		&crsm_sw1,
+		&crsm_sw2,
+		sms_buf_tmp);
+
+	LOG_DBG("EFSMSP: (ret=%d, crsm_sw1=%d, crsm_sw2=%d): %s",
+		ret, crsm_sw1, crsm_sw2, sms_buf_tmp);
+
+	if (ret != 3 || crsm_sw1 != 0x90 || crsm_sw2 != 0x00) {
+		LOG_WRN("Failed to read UICC EFSMSP (ret=%d, crsm_sw1=%d, crsm_sw2=%d): %s",
+			ret, crsm_sw1, crsm_sw2, sms_buf_tmp);
+		goto ta_sim_no_sca;
+	}
+
+	crsm_resp_len = strlen(sms_buf_tmp);
+	if (crsm_resp_len < CRSM_EFSMSP_STATIC_LEN) {
+		LOG_WRN("CRSM response too short (less than %d): %d",
+			CRSM_EFSMSP_STATIC_LEN, crsm_resp_len);
+		goto ta_sim_no_sca;
+	}
+
+	efsmsp_parameter_indicators_field =
+		sms_buf_tmp + crsm_resp_len - CRSM_EFSMSP_PARAMETER_INDICATORS_FROM_END_INDEX;
+	efsmsp_ts_service_centre_address_field =
+		sms_buf_tmp + crsm_resp_len - CRSM_EFSMSP_TS_SERVICE_CENTRE_ADDRESS_FROM_END_INDEX;
+
+	/* Decode Parameter Indicators */
+	bin_buf_len = hex2bin(efsmsp_parameter_indicators_field, 2, &parameter_indicators, 1);
+	if (bin_buf_len != 1) {
+		LOG_WRN("Invalid Parameter Indicators field");
+		goto ta_sim_no_sca;
+	}
+
+	/* If bit #2 is set, TS-Service Centre Address does not exist */
+	if ((parameter_indicators & 0x02) != 0) {
+		LOG_DBG("No TS-Service Centre Address in SIM");
+		goto ta_sim_no_sca;
+	}
+
+	/* Read TS-Service Centre Address */
+	bin_buf_len = hex2bin(efsmsp_ts_service_centre_address_field, 2, &sca_len, 1);
+
+	if (bin_buf_len != 1) {
+		LOG_ERR("Invalid TS-Service Centre Address length");
+		goto ta_sim_no_sca;
+	}
+
+	if (sca_len == 0) {
+		LOG_ERR("TS-Service Centre Address length is zero");
+		goto ta_sim_no_sca;
+	}
+
+	if (sca_len > 12) {
+		LOG_ERR("TS-Service Centre Address length is too long: %d", sca_len);
+		goto ta_sim_no_sca;
+	}	
+
+	strncpy(sca_field, efsmsp_ts_service_centre_address_field, (sca_len + 1) * 2);
+	LOG_DBG("Setting service center address from SIM (len=%d): %s", sca_len, sca_field);
```

Comment:
Hmmm, we shouldn't need to set this but just exit with empty SCA and modem will handle this as necessary just like for normal live network.

Reply:
This is now removed together with code retrieving the SCA value which we shouldn't need.
Please review.